### PR TITLE
User Balance logic (with some referral logic as well)

### PR DESCRIPTION
**Pull request by yenicelik**
- removed solang-parser and lalrpack from Cargo.lock (these are marked as deprecated, the compiled gives a warning)
- added deposit_contract and deposit_topic to the config/example.toml
- added config/development_polygon.toml (to save some work adding RPC providers, feel free to delete this in a future revision / at this point, i can post it here as a comment)
- Decimal digits before and after the comma are specified (20 before, 10 after for now)
- added brownie test suite to simplify reproduction of depositing to payment contracts. Contracts are not needed as the data is read from polygonscan
- put payment deposit logic into frontend/payment.rs . withdrawal will also be more complex (most likely), so a separate file is probably justified
- added logic to compute cost from inbound bytes, outbound bytes, and whether or not the response result is in the cache

### CODE REVIEW

**Comment by yenicelik** [comment]
TODO to David: Clean up files (payment.rs and stats/mod.rs, the order of operations / variable re-use, database re-use, surplus comments & TODOs)

File: web3_proxy/src/rpcs/request.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ pub enum OpenRequestResult {
 
 /// Make RPC requests through this handle and drop it when you are done.
 /// Opening this handle checks rate limits. Developers, try to keep opening a handle and using it as close together as possible
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (yenicelik) [inline web3_proxy/src/rpcs/request.rs:30]: TODO, not sure if this is a good idea, this was needed for the handle, could be a good idea to refactor as well

File: web3_proxy/src/frontend/payments.rs [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    //     _ => Err(Web3ProxyError::BadRequest(
+    //         "Transaction Hash was not found!".to_owned(),
+    //     )),
+    // }?;
+    // warn!("Accepted tokens are: {:?}", accepted_token);
+
+    let (transaction_receipt, accepted_token, decimals): (TransactionReceipt, Address, u32) =
+        match app
+            .balanced_rpcs
+            .best_available_rpc(&authorization, None, &[], None, None)
```
Comment (yenicelik) [inline web3_proxy/src/frontend/payments.rs:182]: @Bryan your feedback here would be nice!

File: web3_proxy/src/rpcs/request.rs (hunk 1 above)
Comment (BlinkyStitt) [reply web3_proxy/src/rpcs/request.rs:30]: I wouldn't be too afraid of clones. Once it works we can do some profiling and make some flamegraphs and see if it's slow. A lot of times the compiler is smart about optimizing clones away

**Comment by yenicelik** [comment]
included in https://github.com/llamanodes/web3-proxy/pull/44 so i will close this

**Pull request closed without merging**